### ISSUE TITLE
fatal: /run belongs to uid 0 instead of 999

### ISSUE THREAD

**Issue by rauanmayemir** [just-containers/s6-overlay#600]
Bumping the image from `v3.1.6.2` to `v3.2.0.2` resulted in containers failing with:
```
/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 999 and we're lacking the privileges to fix it.
s6-overlay-suexec: fatal: child failed with exit code 100
```

I'm running the pod under a non-privileged uid 999 with read-only root fs, relevant dirs have all the proper chowns, everything worked fine before the upgrade.

**Comment by skarnet** [comment]
Do you have a `/run` on your image? How is it mounted?

**Comment by rauanmayemir** [comment]
I do have it, it's an `emptyDir` mount:
```
- mountPath: /run
  name: scratch-volume
  subPath: run
```

Upon building the image, I verified that it's properly owner by my user. But in runtime `/run`'s owner is changed to `root`, even though I've set pod's security context to the least privileged.

**Comment by skarnet** [comment]
OK, two more questions (to inform my diagnosis):

- If your rootfs is read-only, are you running the container with `S6_READ_ONLY_ROOT` set to 1?
- If you run the container with `S6_VERBOSITY` set to 2 or more, early on you should see `preinit` lines with container permissions and `/run` permissions. Can you please reproduce them here?

Thanks.

**Comment by rauanmayemir** [comment]
Yes, I set `S6_READ_ONLY_ROOT` to 1. Will upgrade again tomorrow and collect the details.

**Comment by dkorunic** [comment]
I can confirm that this happens and preinit exits with fatal failure even with `S6_BEHAVIOUR_IF_STAGE2_FAILS=1 set`. An example on K8s that uses `/run` as `emptyDir` same as the above:

```
/package/admin/s6-overlay/libexec/preinit: info: read-only root
/package/admin/s6-overlay/libexec/preinit: info: writable /run. Checking for executability.
/package/admin/s6-overlay/libexec/preinit: info: container permissions: uid=1000 (ubuntu), euid=1000, gid=1000 (ubuntu), egid=1000
/package/admin/s6-overlay/libexec/preinit: info: /run permissions: uid=0 (root), gid=0 (root), perms=oxoworgxgwgruxuwuros
/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 1000 and we're lacking the privileges to fix it.
s6-overlay-suexec: fatal: child failed with exit code 100
```

**Comment by skarnet** [comment]
I see.

So it appears that your container manager mounts `/run` belonging to `root:root` with permissions 1777, just like a `/tmp` temporary filesystem would be mounted.

Unfortunately, this is incorrect. The difference between `/run` and `/tmp` is that *only* root - or, in this case, *only* the uid running the container - should be allowed to create files under `/run`. Doing otherwise opens the system up to denial of service attacks (or, depending on how programs use files under `/run`, prediction attacks, which can be worse).

The current version of s6-overlay detects that the `/run` permissions are incorrect and refuses to run. Previous versions ran, but your system was insecure. I realize this can be perceived as inconvenient, but I believe the change to be an improvement.

The fix is to find where your container manager mounts `/run` and change how it is mounted. It should have the uid and gid of the process that runs the container, and have 0755 permissions.

If that proves too difficult, I can probably add an option in a future version of s6-overlay that allows it to run on insecure `/run` configurations, but that won't be right away, and that won't be a solution, just a dirty workaround. Please try and fix your container manager instead.

**Comment by dkorunic** [comment]
@skarnet At least in K8s world, sadly it's not possible to change `emptyDir` ownership when running in rootless container. Directory will be created in-memory with UID 0 owner and the only thing we can change is group ownership through pod-level `fsGroup` field (as it can be seen below with `gid=1000` for `/run`), but `preinit` still throws a fatal error:

```
/package/admin/s6-overlay/libexec/preinit: info: container permissions: uid=1000 (ubuntu), euid=1000, gid=1000 (ubuntu), egid=1000
/package/admin/s6-overlay/libexec/preinit: info: /run permissions: uid=0 (root), gid=1000 (ubuntu), perms=oxoworgxgwgruxuwurosgs
/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 1000 and we're lacking the privileges to fix it.
```

Since we use s6 extensively in K8s, it would be great if we could silence this or just have it as non-fatal warning, since as long as container runs as non-UID 0, everything inside is also non-root (and mapped to a single user) and `/run` tmpfs mount security is really not relevant, given all processes belong to the same UID anyway.

**Comment by skarnet** [comment]
It is true that the mapping to a single uid makes the `/run` permissions less problematic. I'll see what support I can add to this use case - but it is *still* a defect in k8s that is worth reporting. `/run` should be mounted as the container user, always.

**Comment by skarnet** [comment]
If you have the ability to build s6-overlay yourself, could you please test the latest commit? It should print a warning, but still run your container. If it works for you, I'll cut a new release.

**Comment by dkorunic** [comment]
@skarnet I can confirm that build from the latest tip (https://github.com/just-containers/s6-overlay/commit/42e902133b95307ef5cebad575a254bb18f456a8 and https://github.com/just-containers/s6-overlay/commit/1705b0f022ecbdb67663e5d5160b35b46f609328) prints a warning but proceed booting as expected, as seen below:

```
/package/admin/s6-overlay/libexec/preinit: info: read-only root
/package/admin/s6-overlay/libexec/preinit: info: writable /run. Checking for executability.
/package/admin/s6-overlay/libexec/preinit: warning: /run belongs to uid 0 instead of 1000, but we can still work in single-uid mapping.
```

Thank you! I can confirm that this helps.

**Comment by taka-oyama** [comment]
For anyone interested, an alternate way to make this work is to change the permission of `/run` using [initContainer](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/).
```yaml
initContainers:
- name: setup-s6-overlay
  image: YOUR_IMAGE
  securityContext:
    privileged: true
  volumeMounts:
  - name: run-volume
    mountPath: /run
  command:
  - sh
  - -c
  - chown 1000:1000 /run
```

**Comment by dkorunic** [comment]
@taka-oyama Absolutely, it has been mentioned [already](https://github.com/just-containers/s6-overlay/issues/600#issuecomment-2537983015), but the main issue with privileged initContainers is the lack of restrictions and with some K8s providers (for instance GKE AutoPilot) this won't work at all. Permitting any kind of container to run privileged allows quite a few vectors of attack resulting in ServiceAccounts compromise and even host/hypervisor compromise.

**Comment by taka-oyama** [comment]
> Absolutely, it has been mentioned https://github.com/just-containers/s6-overlay/issues/600#issuecomment-2537983015

Oof, that's Embarrassing. I must have skimmed through that part. 

> Permitting any kind of container to run privileged allows quite a few vectors of attack resulting in ServiceAccounts compromise and even host/hypervisor compromise.

Thank you for the enlightenment. I thought it was ok because it's only running in privileged mode before the pod became ready.

**Comment by dkorunic** [comment]
> Thank you for the enlightenment. I thought it was ok because it's only running in privileged mode before the pod became ready.

Yes, but look from the K8s admin perspective -- a single image with possibly unverified content is able to run one or more commands in fully privileged mode. Those commands (`sh` for instance) can be altered to execute sandbox escape, for instance load eBPF module in the host kernel. And from there it's game over: attackers can "own" ServiceAccounts, create new Pods (even fully privileged), change and terminate processes on other Pods, create hidden VPNs (eBPF XDP/TC) and pretty much do anything as root on the host.
That's why that majority of the K8s security assessment tools will warn about any kind of privileged containers, as they pose a significant risk to the infrastructure as whole.

**Comment by taka-oyama** [comment]
Really appreciate the detailed explanation. Thank you. 
I remove the code and use the fix provided.

**Comment by rauanmayemir** [comment]
Could a new release be tagged with a fix for this please?

**Comment by skarnet** [comment]
Yeah, sorry it's being slow. I'm currently investigating a problem report which I *believe* has nothing to do with s6-overlay, but I would like to see it fixed before tagging a new release, just in case.

**Comment by dkorunic** [comment]
Sorry to be a bother, but this is still an issue (with a perfectly legit K8s deployment).

**Comment by skarnet** [comment]
A new release, including the fix, will be coming in the next month. Sorry for not getting there faster, release engineering is a long and work-intensive process and I'm trying to factor the releases for the entire software stack.

**Comment by fgeck** [comment]
Sorry to bother again here - but Is there any rough eta for the next release?

**Comment by skarnet** [comment]
@rauanmayemir You never answered the second question I asked [here](https://github.com/just-containers/s6-overlay/issues/600#issuecomment-2508071669), so I assumed you were happy with the analyses from @dkorunic.

If your use case is not addressed by the change that fixes @dkorunic's case, please give all the debug information that is relevant to your own case.

**Comment by rauanmayemir** [comment]
@skarnet sorry about that, I thought I was having same issues. Here are my logs with `S6_VERBOSITY=3` (2 and 3 have no difference)

```
/package/admin/s6-overlay/libexec/preinit: info: read-only root
/package/admin/s6-overlay/libexec/preinit: info: writable /run. Checking for executability.
/package/admin/s6-overlay/libexec/preinit: info: container permissions: uid=65534 (nobody), euid=65534, gid=65534 (nogroup), egid=65534
/package/admin/s6-overlay/libexec/preinit: info: /run permissions: uid=0 (root), gid=65534 (nogroup), perms=oxoworgxgwgruxuwurgs
/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 65534, has insecure permissions, and we're lacking the privileges to fix it.
s6-overlay-suexec: fatal: child failed with exit code 100
```

**Comment by rauanmayemir** [comment]
Kubernetes only allows setting `fsGroup`, so `emptyDir` is owned by `root:nogroup`.

**Comment by rauanmayemir** [comment]
I found that I'm running `chmod 02755 /run` in my Dockerfile, was it supposed to be `chmod 0755 /run`?

**Comment by skarnet** [comment]
`02755` should be fine, but the problem here is that it's **02777**. Your `/run` is world-writable, and since it belongs to `root` but the container is run as `nobody`, s6-overlay can't fix it, and world-writable is just too insecure so it refuses to run.

**Comment by rauanmayemir** [comment]
It's pretty much [hard-coded](https://github.com/kubernetes/kubernetes/blob/893486dfd16ff8c628c6f33bb2bea869ad86115f/pkg/volume/emptydir/empty_dir.go#L50).

**Comment by skarnet** [comment]
This is unfortunate.

There is never a good reason to have `/run` world-writable. Even in a container running as a single non-root user, where *technically* it poses no immediate danger, mounting a world-writable directory as `/run` is a huge footgun waiting to fire the moment someone changes the container permissions, and I don't want s6-overlay to have any part in such a scenario.

Get Kubernetes to fix their stuff.

**Comment by rauanmayemir** [comment]
What's the difference between mounting a world-writeable `/run` and a world-writeable `/any/other/dir`?

Like, right now we've achieved a solution as safe as it supposedly can get with non-privileged containers, read only root fs, disallowed privilege escalation, dropped all capabilities, but all of that turned out meaningless because s6-overlay doesn't like `/run`'s permissions. 🤷‍♂

**Comment by xorinzor** [comment]
It's important to remember that any parent directory permissions do not affect the permissions of the files & directories it contains.

Files inside can have more secure permissions and would still be protected as long as the process attempting to access them does not have those permissions.

**Comment by dkorunic** [comment]
Just to tackle few of these remarks above -- honestly, it would be great if s6 overlay would work out of the box on K8s environments too, whatever mess K8s does. We really can't change `/run` behaviour and if we can't at least disable this failsafe logic in s6 overlay, this means that this otherwise great software is can't be used for K8s environment.

**Comment by skarnet** [comment]
@xorinzor And the whole subdirectory can still be renamed or deleted by anyone, so, yeah, nice protection you've got there.

@rauanmayemir There's no difference. A world-writable directory is a terrible idea *anywhere*. But `/run` is a widely accepted convention for a directory where **root** stores its information, and it's supposed to be reliable and non-falsifiable by non-root users. Allowing `/run` to *not* be root-owned is already a stretch, a hack to make user containers work; it's fine as long as it behaves the same way and we can pretend it's still a somewhat regular Linux machine. But having `/run` world-writable is another level of not caring and it breaks that suspension of disbelief, you know? And I find it rather unpleasant that you are choosing to blame s6-overlay for it, instead of, oh I don't know, Kubernetes. Especially since you had *months* to answer a question and make me aware of the issue.

@dkorunic I suppose it's what's going to have to happen, but *why* is it so difficult to get Kubernetes to do the right thing, and *why* are you insisting that *I* change stuff to make things *less* secure instead of insisting that *they* change stuff to make things *more* secure?

**Comment by tlemarchand** [comment]
> `02755` should be fine, but the problem here is that it's **02777**. Your `/run` is world-writable, and since it belongs to `root` but the container is run as `nobody`, s6-overlay can't fix it, and world-writable is just too insecure so it refuses to run.

While I appreciate your security concerns about the 02777 permission, I respectfully disagree with S6's approach here.
In container environments, the responsibility for security posture should ultimately rest with the user/administrator who is deploying and configuring the environment. S6-overlay is a process supervision system, not a security enforcement tool. By refusing to run in this scenario, I think that S6 is overstepping its core functionality and imposing security decisions that should be left to the operators.

Additionally, in Kubernetes environments, workloads are already isolated from each other through the platform's security mechanisms. The container runtime provides namespace isolation, and Kubernetes adds additional security boundaries between pods. Given this isolation, a world-writable /run directory within a container doesn't present the same security risks as it would in a traditional shared system. The container boundary itself prevents other workloads from accessing this directory, making S6's strict permission enforcement unnecessary and potentially disruptive to valid use cases.

You already log an error to raise awareness in potential cases where it would be exploitable.
It seems to me that it's a good and sufficient option.

**Comment by rauanmayemir** [comment]
Don’t get me wrong, your rigorous approach is what I love about s6. I’m just pleading to consider practical aspects.

**Comment by skarnet** [comment]
If you can build s6-overlay yourself, please test with the latest commit. Add
`S6_YES_I_WANT_A_WORLD_WRITABLE_RUN_BECAUSE_KUBERNETES=1` to your environment, and you should be able to run s6-overlay even with these wacko permissions.

No, I will not be changing the name of the variable.

**Comment by skarnet** [comment]
@xorinzor Try it yourself:
```
root # mkdir -p -m 0777 /var/test && chown nobody:nobody /var/test  # or don't chown, it changes nothing

huh $ mkdir -m 0700 /var/test/huh
huh $ echo "huh's preciousss data" > /var/test/huh/data

lol $ rm -rf /var/test/huh
lol $ ls -l /var/test
total 0

huh $ echo 'huh?'
huh?

lol $ echo 'lol'
lol
```

Yes, I understand that k8s pods are supposed to be isolated, yadda yadda, but nevertheless, using world-writable directories in Unix is a *pretty fucking stupid* thing to do, and even if they're never supposed to be exploitable, it just shows that the people who made that choice have never heard of defense in depth, or never considered the possibility that their creation could one day be used in a situation that they might not have planned for or control. Which, in either case, screams *inexperience*.

**Comment by xorinzor** [comment]
@skarnet hm interesting, perhaps I'm misunderstanding what I read about the permission system, will look that up again sometime.

**Comment by tlemarchand** [comment]
> [@skarnet](https://github.com/skarnet) hm interesting, perhaps I'm misunderstanding what I read about the permission system, will look that up again sometime.

The behavior that you expect is enabled by using a sticky bit on the folder : https://en.wikipedia.org/wiki/Sticky_bit

There is currently a PR in Kubernetes to be able to set the sticky bit on an emptyDir volume mount : https://github.com/kubernetes/kubernetes/pull/130277
This will limit the security risk compared to 02777.

**Comment by skarnet** [comment]
The stickyb it is the correct semantics for `/tmp`, but not for `/run`, so this PR wouldn't be good enough here, but it would improve the situation *somewhat*.

If there's a PR for the sticky bit, I don't understand why there cannot be a PR so that a directory can be mounted with `0700` or `2755` permissions, though. Why is doing the real right thing so difficult? 🤦

**Comment by rauanmayemir** [comment]
@skarnet I checked and this works, thank you! 🙌 

P.S: last time I tried building from sources, I wasted most of the day and still failed to have a proper build. I just patched the `preinit` script instead.

**Comment by skarnet** [comment]
All right, new 3.2.1.0 release includes the change. Closing this issue; if there's _still_ a case that's not covered, please reopen.

**Comment by michaelbeaumont** [comment]
So it's not working for me still with `S6_YES_I_WANT_A_WORLD_WRITABLE_RUN_BECAUSE_KUBERNETES=1`, but I have:

```
/package/admin/s6-overlay/libexec/preinit: info: container permissions: uid=1000 (paperless), euid=1000, gid=1000 (paperless), egid=1000
/package/admin/s6-overlay/libexec/preinit: info: /run permissions: uid=0 (root), gid=0 (root), perms=oxorgxgruxuwur
/package/admin/s6-overlay/libexec/preinit: fatal: /run belongs to uid 0 instead of 1000, has insecure and/or unworkable permissions, and we're lacking the privileges to fix it.
```

It seems like the check is rather "simple":

```
s6-grep -q "oxoworgxgwgruxuwur"
```

is there a reason the permissions have to match exactly? There are a number of Kubernetes options such that the permissions will be slightly different. Can we just skip this check?

**Comment by michaelbeaumont** [comment]
For anyone else who has the same issue as me, the "fix" is to set `securityContext.readOnlyRootFilesystem: true` (and the `S6_READ_ONLY_ROOT=1` env var)

**Comment by skarnet** [comment]
This is generally a good idea anyway!

**Comment by michaelbeaumont** [comment]
> You obviously need an emptyDir mounted on /run.

You don't and I'm not sure why you would. You just need to set the two environment variables and have `readOnlyRootFilesystem: true` when using s6, apparently.

**Comment by tlemarchand** [comment]
> > You obviously need an emptyDir mounted on /run.
> 
> You don't and I'm not sure why you would. You just need to set the two environment variables and have `readOnlyRootFilesystem: true` when using s6, apparently.

This is what you get without /run :

```
/package/admin/s6-overlay/libexec/preinit: info: read-only root                                                                              
/package/admin/s6-overlay-3.2.1.0/libexec/preinit: 41: cannot create /run/test of writability: Read-only file system                         
s6-overlay-suexec: fatal: child failed with exit code 2    
```

Maybe your configuration does not really enforce read-only ?

**Comment by rauanmayemir** [comment]
@michaelbeaumont If you don’t mount emptyDir, then your /run will not be writeable. Something might break.
